bpf: Fix 4.19 program size and complexity issues

pchaigno opened the pull request:
This PR does mainly two things, implemented by the 10 following commits.

#### Fix program size and complexity issues:

- Enforce host policies before BPF-based SNAT to allow for following commit fix.
- Add tail call to fix program size issue.
- Simplify and extend `tail_call.cocci`.
- Remove `relax_verifier()` that increases complexity.
- Tail call even if IPv6 is disabled to fix complexity issue.

#### Enable `kube-proxy-replacement=strict` in 4.19 CI build:

- Disable host firewall if BPF-based egress LB is enabled.
- Add `RunsOn419Kernel()` helper
- Helm option for fragment tracking to be able to disable it in CI.
- Disable fragment tracking in 4.19 CI.
- Finally: disable kube-proxy in 4.19 CI and use `kube-proxy-replacement=strict`.

Fixes: #11175.

/cc @brb

pchaigno committed:
bpf: Enforce host policies before BPF-based SNAT

This fix is required to get the host policies to work properly with
BPF-backed NodePort services. It is not a sufficient fix, but required
to resolve the 4.19 program size issue in subsequent commits, so sending
now.

Fixes: 88bf291 ("bpf: Enforce host policies for IPv4")
Signed-off-by: Paul Chaignon <paul@cilium.io>

pchaigno committed:
daemon: Disable host firewall if BPF LB is enabled

Signed-off-by: Paul Chaignon <paul@cilium.io>

qmonnet commented on contrib/coccinelle/tail_calls.cocci:
```diff
@@ -30,50 +30,58 @@ symbol ret;
 |
   // We also whitelist any function returning DROP_MISSED_TAIL_CALL, assuming
-  // this will be catch afterwards and transformed in call to
+  // this will be catched afterwards and transformed in call to
```
“caught”

qmonnet commented on bpf/lib/nodeport.h:
```diff
@@ -799,11 +812,12 @@ static __always_inline bool nodeport_nat_ipv4_needed(struct __ctx_buff *ctx,
-		if (encap)
+		 * (IS_BPF_OVERLAY denotes this fact). Otherwise, a packet will
+		 * be SNAT'd to cilium_host IP addr. */
 			return false;
```
Nit: indent

qmonnet commented on install/kubernetes/quick-install.yaml:
```diff
@@ -164,6 +164,7 @@ data:
   enable-xt-socket-fallback: "true"
   install-iptables-rules: "true"
   auto-direct-node-routes: "false"
+  enable-ipv4-fragment-tracking: "true"
```
Is this necessary? The option should default to `true` if not provided?

pchaigno replied:
That file was generated by `make`. I think those are all the default values.

joestringer replied:
I think if you want it to be omitted here then you need to change the configmap to have `{{ -if ... }}` logic wrapped around the option. (I don't have any opinion that this would help or hinder users)

pchaigno replied:
Yep, I followed the pattern of other options. I kinda makes sense to me that we'd want to list all options here. It makes it easier for the user to know whether the `quick-install.yml` will enable e.g., fragment tracking.

qmonnet commented on test/k8sT/Services.go:
```diff
@@ -1330,7 +1330,16 @@ var _ = Describe("K8sServicesTest", func() {
+				options["global.fragmentTracking"] = "true"
+				options["global.debug.enabled"] = "false"
+			}
+			DeployCiliumOptionsAndDNS(kubectl, ciliumFilename, options)
```
Nit: Even if it defaults to `true`, we could maybe set `options["global.fragmentTracking"]` to `true` explicitly for all kernels for this test? (I should have done that and the Helm option, my bad.)
```suggestion
			var options map[string]string
			// On 4.19, because we're running with
			// kube-proxy-replacement=strict, we need to choose between
			// debug mode and fragment tracking support to avoid complexity
			// issues.
			options["global.fragmentTracking"] = "true"
			options["global.debug.enabled"] = !helpers.RunsOn419Kernel()
			DeployCiliumOptionsAndDNS(kubectl, ciliumFilename, options)
```

qmonnet reviewed (commented):
All changes so far look good to me (minor nits inline below), thanks a lot for this work!

joestringer commented on bpf/lib/common.h:
```diff
@@ -71,12 +71,13 @@
+#define CILIUM_CALL_IPV6_NODEPORT_DSR		22
+#define CILIUM_CALL_IPV4_FROM_HOST		23
+#define CILIUM_CALL_IPV6_FROM_HOST		24
+#define CILIUM_CALL_SIZE			25
```
The corresponding change to the verifier complexity script wasn't included in this commit, was it later or somehow omitted?

pchaigno replied:
Omitted but not forgotten. I'm just checking complexity issues so far, hence the draft mode. I also suspect these few line changes are the reason for the downgrade test failure on net-next.

joestringer commented on test/helpers/utils.go:
```diff
@@ -497,11 +497,21 @@ func DoesNotRunOnNetNextKernel() bool {
+	return os.Getenv("KERNEL") == "419"
+}
+
+// DoesNotRunOn419Kernel is the complement function of RunsOn419Kernel.
+func DoesNotRunOn419Kernel() bool {
+	return !RunsOn419Kernel()
+}
```
nit: Can't we just open-code the `!RunsOn419Kernel()` in the callers?

pchaigno replied:
Yeah, probably. I was following the pattern established by other, similar helpers.

pchaigno committed:
bpf: Fix 4.19 complexity issue when IPv6 is disabled

When IPv6 is disabled (e.g., default quick-install), the verifier rejects
the from-container section of bpf_lxc with a complexity error. This
error happens because when one of IPv6 and IPv4 is disabled, we don't
use a tail call to call tail_handle_ipv{4,6} and from-container gets
just long enough to cause a complexity error.

This commit fixes it by always using a tail call to execute
tail_handle_ipv{4,6}.

Signed-off-by: Paul Chaignon <paul@cilium.io>

pchaigno committed:
test: Disable fragment tracking support in 4.19 CI build

We disable it to avoid hitting complexity issues on that kernel. The
fragment tracking test disables debug mode instead.

Signed-off-by: Paul Chaignon <paul@cilium.io>

pchaigno committed:
bpf: Fix program size issue on 4.19

On Linux 4.19 (and any kernel with 4096 BPF instructions limit), when
running with kube-proxy-replacement=strict, IPv4, IPv6, fragment support
and in debug mode, Cilium fails to load bpf_overlay and bpf_host with:

  Prog section '2/19' rejected: Argument list too long (7)
   - Type:         3
   - Attach Type:  0
   - Instructions: 4203 (107 over limit)
   - License:      GPL
  Error filling program arrays!

2/19 is section CILIUM_CALL_NODEPORT_REVNAT. We can reduce the number of
instructions in that section by breaking the IPv4 and IPv6 cases with
tail calls. Instead of introducing a new tail call on each path, we can
turn CILIUM_CALL_NODEPORT_REVNAT into CILIUM_CALL_IPV{4,6}_NODEPORT_REVNAT.

Signed-off-by: Paul Chaignon <paul@cilium.io>

pchaigno committed:
cocci: Simplify and extend tail_call.cocci

This commit extends tail_call.cocci to:
- recognize invoke_tailcall_if() calls followed by 'return ret;', as
  used in nodeport.h;
- refactor the code a bit using disjunctions;
- ensure the return statement we are trying to match (e.g., return
  DROP_MISSED_TAIL_CALL or return send_drop_notify()) is not preceeded by
  another return statement.

Fixes: f02bd2a ("cocci: Detect unlogged missed tail calls")
Signed-off-by: Paul Chaignon <paul@cilium.io>

pchaigno committed:
ci: Disable kube-proxy in 4.19 CI build

Signed-off-by: Paul Chaignon <paul@cilium.io>

pchaigno committed:
bpf: Remove relax_verifier() in CILIUM_CALL_IPV6_FROM_LXC

relax_verifier() is supposed to help the verifier prune branches and
therefore decrease the reported complexity. On 4.19 however, the call to
relax_verifier in CILIUM_CALL_IPV6_FROM_LXC actually increases the
complexity because it adds instructions in complexity-hot paths.

After a quick check, it doesn't seem to be required for 4.9 either, so
let's see if we can remove it.

Signed-off-by: Paul Chaignon <paul@cilium.io>

pchaigno committed:
test: RunsOn419Kernel() helper

Signed-off-by: Paul Chaignon <paul@cilium.io>

pchaigno committed:
install/kubernetes: Helm option for fragment tracking

Signed-off-by: Paul Chaignon <paul@cilium.io>
